include/mujoco/mujoco.h

```diff
@@ -103,43 +103,54 @@ MJAPI void mj_deleteVFS(mjVFS* vfs);
+MJAPI mjModel* mj_loadXML(const char* NONNULL_ARG filename,
+                          const mjVFS* NULLABLE_ARG vfs,
+                          char* NULLABLE_ARG error, int error_sz)
+    NONNULL_FUNC(1);
 
 // Update XML data structures with info from low-level model, save as MJCF.
 // If error is not NULL, it must have size error_sz.
-MJAPI int mj_saveLastXML(const char* filename, const mjModel* m, char* error, int error_sz);
+MJAPI int mj_saveLastXML(const char* NONNULL_ARG filename,
+                         const mjModel* NONNULL_ARG m, char* NULLABLE_ARG error,
+                         int error_sz) NONNULL_FUNC(1, 2);
 
 // Free last XML model if loaded. Called internally at each load.
 MJAPI void mj_freeLastXML(void);
 
 // Print internal XML schema as plain text or HTML, with style-padding or &nbsp;.
-MJAPI int mj_printSchema(const char* filename, char* buffer, int buffer_sz,
-                         int flg_html, int flg_pad);
-
+MJAPI int mj_printSchema(const char* NULLABLE_ARG filename,
+                         char* NULLABLE_ARG buffer, int buffer_sz, int flg_html,
+                         int flg_pad);
 
 //---------------------------------- Main simulation -----------------------------------------------
 
 // Advance simulation, use control callback to obtain external force and control.
-MJAPI void mj_step(const mjModel* m, mjData* d);
+MJAPI void mj_step(const mjModel* NONNULL_ARG m, mjData* NONNULL_ARG d)
+    NONNULL_FUNC(1, 2);
 
 // Advance simulation in two steps: before external force and control is set by user.
-MJAPI void mj_step1(const mjModel* m, mjData* d);
+MJAPI void mj_step1(const mjModel* NONNULL_ARG m, mjData* NONNULL_ARG d)
+    NONNULL_FUNC(1, 2);
 
 // Advance simulation in two steps: after external force and control is set by user.
-MJAPI void mj_step2(const mjModel* m, mjData* d);
+MJAPI void mj_step2(const mjModel* NONNULL_ARG m, mjData* NONNULL_ARG d)
+    NONNULL_FUNC(1, 2);
 
 // Forward dynamics: same as mj_step but do not integrate in time.
-MJAPI void mj_forward(const mjModel* m, mjData* d);
+MJAPI void mj_forward(const mjModel* NONNULL_ARG m, mjData* NONNULL_ARG d)
+    NONNULL_FUNC(1, 2);
 
 // Inverse dynamics: qacc must be set before calling.
-MJAPI void mj_inverse(const mjModel* m, mjData* d);
+MJAPI void mj_inverse(const mjModel* NONNULL_ARG m, mjData* NONNULL_ARG d)
+    NONNULL_FUNC(1, 2);
 
 // Forward dynamics with skip; skipstage is mjtStage.
-MJAPI void mj_forwardSkip(const mjModel* m, mjData* d, int skipstage, int skipsensor);
+MJAPI void mj_forwardSkip(const mjModel* NONNULL_ARG m, mjData* NONNULL_ARG d,
+                          int skipstage, int skipsensor) NONNULL_FUNC(1, 2);
 
 // Inverse dynamics with skip; skipstage is mjtStage.
-MJAPI void mj_inverseSkip(const mjModel* m, mjData* d, int skipstage, int skipsensor);
-
+MJAPI void mj_inverseSkip(const mjModel* NONNULL_ARG m, mjData* NONNULL_ARG d,
+                          int skipstage, int skipsensor) NONNULL_FUNC(1, 2);
```

Comment:
after running -

```elixir
➜  mujoco git:(309-Adding-_Nonull-and-_Nullable-attributes-to-mujoco.h) git clang-format HEAD~1 
changed files:
    include/mujoco/mjmacro.h
    include/mujoco/mujoco.h
➜  mujoco git:(309-Adding-_Nonull-and-_Nullable-attributes-to-mujoco.h) git clang-format HEAD~1 
changed files:
    include/mujoco/mujoco.h
```

The asterisk * is becoming insistent with the rest of the code for the pointers in the updated definitions.

Moreover I am looking at the conflict which is going with the mujoco.h file.

Reply:
Is this the conflict between a version of the file that uses the NONNULL_ARG and NONNULL_FUNC macros, and a version that does not use those macros?

Reply:
I will look into the changes with the comments in the later part of the mujoco.h file in the morning.

The changes I made are just before the ray-collisions. =)